File: src/main/java/org/eln/eln3/sim/process/heater/DiodeHeatThermalLoad.java
package org.eln.eln3.sim.process.heater;

import org.eln.eln3.sim.IProcess;
import org.eln.eln3.sim.ThermalLoad;
import org.eln.eln3.sim.mna.component.Resistor;

public class DiodeHeatThermalLoad implements IProcess {

    Resistor resistor;
    ThermalLoad load;
    double lastResistance;

    public DiodeHeatThermalLoad(Resistor r, ThermalLoad load) {
        this.resistor = r;
        this.load = load;
        lastResistance = r.getResistance();
    }

    @Override
    public void process(double time) {
        if (resistor.getResistance() == lastResistance) {
            load.movePowerTo(resistor.getPower());
        } else {
            lastResistance = resistor.getResistance();
        }
    }
}

File: src/main/kotlin/org/eln/eln3/sim/process/thermal/DiodeHeatThermalLoad.kt
package org.eln.eln3.sim.process.thermal

import org.eln.eln3.sim.IProcess
import org.eln.eln3.sim.ThermalLoad
import org.eln.eln3.sim.mna.component.Resistor

class DiodeHeatThermalLoad(var resistor: Resistor, var load: ThermalLoad) : IProcess {
    var lastResistance: Double

    init {
        lastResistance = resistor.resistance
    }

    override fun process(time: Double) {
        if (resistor.resistance == lastResistance) {
            load.addThermalPower(resistor.power)
        } else {
            lastResistance = resistor.resistance
        }
    }
}